src/sparse_blas/backends/rocsparse/rocsparse_handles.cpp

```diff
@@ -0,0 +1,491 @@
+
+#include "oneapi/mkl/sparse_blas/detail/rocsparse/onemkl_sparse_blas_rocsparse.hpp"
+
+#include "rocsparse_error.hpp"
+#include "rocsparse_helper.hpp"
+#include "rocsparse_handles.hpp"
+#include "rocsparse_scope_handle.hpp"
+#include "rocsparse_task.hpp"
+#include "sparse_blas/macros.hpp"
+
+namespace oneapi::mkl::sparse::rocsparse {
+
+/**
+ * In this file RocsparseScopedContextHandler are used to ensure that a rocsparse_handle is created before any other rocSPARSE call, as required by the specification.
+*/
+
+// Dense vector
+template <typename fpType>
+void init_dense_vector(sycl::queue& queue, dense_vector_handle_t* p_dvhandle, std::int64_t size,
+                       sycl::buffer<fpType, 1> val) {
+    auto event = queue.submit([&](sycl::handler& cgh) {
+        auto acc = val.template get_access<sycl::access::mode::read_write>(cgh);
+        detail::submit_host_task(cgh, queue, [=](sycl::interop_handle ih) {
+            // Ensure that a rocsparse handle is created before any other rocSPARSE function is called.
+            detail::RocsparseScopedContextHandler(queue, ih).get_handle(queue);
+            auto roc_value_type = detail::RocEnumType<fpType>::value;
+            rocsparse_dnvec_descr roc_dvhandle;
+            ROCSPARSE_ERR_FUNC(rocsparse_create_dnvec_descr, &roc_dvhandle, size,
+                               detail::get_mem(ih, acc), roc_value_type);
+            *p_dvhandle = new dense_vector_handle(roc_dvhandle, val, size);
+        });
+    });
+    event.wait_and_throw();
+}
+
+template <typename fpType>
+void init_dense_vector(sycl::queue& queue, dense_vector_handle_t* p_dvhandle, std::int64_t size,
+                       fpType* val) {
+    auto event = queue.submit([&](sycl::handler& cgh) {
+        detail::submit_host_task(cgh, queue, [=](sycl::interop_handle ih) {
+            // Ensure that a rocsparse handle is created before any other rocSPARSE function is called.
+            detail::RocsparseScopedContextHandler(queue, ih).get_handle(queue);
+            auto roc_value_type = detail::RocEnumType<fpType>::value;
+            rocsparse_dnvec_descr roc_dvhandle;
+            ROCSPARSE_ERR_FUNC(rocsparse_create_dnvec_descr, &roc_dvhandle, size, val,
+                               roc_value_type);
+            *p_dvhandle = new dense_vector_handle(roc_dvhandle, val, size);
+        });
+    });
+    event.wait_and_throw();
+}
+
+template <typename fpType>
+void set_dense_vector_data(sycl::queue& queue, dense_vector_handle_t dvhandle, std::int64_t size,
+                           sycl::buffer<fpType, 1> val) {
+    detail::check_can_reset_value_handle<fpType>(__func__, dvhandle, true);
+    auto event = queue.submit([&](sycl::handler& cgh) {
+        auto acc = val.template get_access<sycl::access::mode::read_write>(cgh);
+        detail::submit_host_task(cgh, queue, [=](sycl::interop_handle ih) {
+            if (dvhandle->size != size) {
```

Comment:
I wonder if we should throw an exception if `dvhandle->size < size` instead. E.g., users may want to pad the`sycl::buffer` for some reason with zeros in the end, which wouldn't work with the current `if` condition.

Reply:
Nevermind, this is `set_dense_vector_data` used for replacing/switching out arrays in an existing handle. The condition is fine then.

Reply:
Although maybe it doesn't hurt to add the check both here and in `init_xxx_data`

Reply:
The check to verify that the buffer size is compatible with the handle size is done elsewhere, see https://github.com/oneapi-src/oneMKL/blob/develop/src/sparse_blas/generic_container.hpp#L164
This is a common place that the other backends also use.